browser/ui/views/brave_actions/brave_actions_container.cc

```diff
@@ -0,0 +1,195 @@
+#include "extensions/browser/extension_registry_observer.h"
+#include "extensions/common/one_shot_event.h"
+
+BraveActionsContainer::BraveActionsContainer(Browser* browser, Profile* profile)
+  : views::View(),
+    browser_(browser),
+    extension_action_api_(extensions::ExtensionActionAPI::Get(profile)),
+    extension_registry_(extensions::ExtensionRegistry::Get(profile)),
+    extension_action_manager_(
+      extensions::ExtensionActionManager::Get(profile)
+    ),
+    extension_registry_observer_(this),
+    extension_action_observer_(this),
+    weak_ptr_factory_(this) {
+  // Handle when the extension system is ready
+  extensions::ExtensionSystem::Get(profile)->ready().Post(
+      FROM_HERE, base::Bind(&BraveActionsContainer::OnExtensionSystemReady,
+                            weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveActionsContainer::~BraveActionsContainer() {}
+
+void BraveActionsContainer::Init() {
+  // automatic layout
+  auto vertical_container_layout = std::make_unique<views::BoxLayout>(
+                                                views::BoxLayout::kHorizontal);
+  vertical_container_layout->set_main_axis_alignment(
+                                  views::BoxLayout::MAIN_AXIS_ALIGNMENT_CENTER);
+  vertical_container_layout->set_cross_axis_alignment(views::BoxLayout::CROSS_AXIS_ALIGNMENT_CENTER);
+  SetLayoutManager(std::move(vertical_container_layout));
+
+  // children
+  views::Separator* brave_button_separator_ = new views::Separator();
+  // TODO: theme color
+  brave_button_separator_->SetColor(SkColorSetRGB(0xb2, 0xb5, 0xb7));
+  brave_button_separator_->SetPreferredSize(gfx::Size(2,
+                            GetLayoutConstant(LOCATION_BAR_ICON_SIZE)));
+  AddChildView(brave_button_separator_);
+}
+
+void BraveActionsContainer::AddShields(
+                                const extensions::Extension* brave_extension) {
+  DCHECK(brave_extension);
+  VLOG(1) << "AddShields, was already loaded:" << (bool)shields_button_view_;
+  if (!shields_button_view_) {
+    // Create a ExtensionActionViewController for the extension
+    // Passing |nullptr| instead of ToolbarActionsBar since we do not require that
+    // logic.
+    // If we do require notifications when popups are open or closed,
+    // then we should inherit and pass |this| through.
+    shields_view_controller_ = new ShieldsActionViewController(brave_extension,
+        browser_, extension_action_manager_->GetExtensionAction(*brave_extension),
+                                                                        nullptr);
+    shields_button_view_ = new ToolbarActionView(
+                                                  shields_view_controller_, this);
+    // Sets overall size of button but not image graphic. We set a large width
+    // in order to give space for the bubble.
+    shields_button_view_->SetPreferredSize(gfx::Size(32, 24));
+    // Add extension view after separator view
+    AddChildViewAt(shields_button_view_, 1);
```

Comment:
`AddChildView()` adds newly added child to last.
```
void View::AddChildView(View* view) {
  if (view->parent_ == this)
    return;
  AddChildViewAt(view, child_count());
}
```

Reply:
Yes. I had made this explicit as index 1 since we will be adding another button which we will want at index 2. In other words the order of the buttons must always be the same. But yes, when we do add that button in a future PR, we won't be able to add it at index 2 if the shields hasn't loaded yet to index 1, so we'll need a better way of positioning.